eland/ml/pytorch/pytorch_model.py

```diff
@@ -0,0 +1,154 @@
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+import base64
+import json
+import math
+import os
+from typing import TYPE_CHECKING, Any, Dict, List, Set, Tuple, Union
+
+import elasticsearch
+from tqdm.auto import tqdm
+
+from eland.common import ensure_es_client
+
+if TYPE_CHECKING:
+    from elasticsearch import Elasticsearch
+
+DEFAULT_CHUNK_SIZE = 4 * 1024 * 1024  # 4MB
+QUEUE_SIZE = 4
+THREAD_COUNT = 4
+DEFAULT_TIMEOUT = "60s"
+
+
+class PyTorchModel:
+    """
+    A PyTorch model managed by Elasticsearch.
+
+    These models must be trained outside of Elasticsearch, conform to the
+    support tokenization and inference interfaces, and exported as their
+    TorchScript representations.
+    """
+
+    def __init__(
+        self,
+        es_client: Union[str, List[str], Tuple[str, ...], "Elasticsearch"],
+        model_id: str,
+    ):
+        self._client = ensure_es_client(es_client)
+        self.model_id = model_id
+
+    @staticmethod
+    def _load_json(path: str) -> Dict[str, Any]:
+        with open(path, "r") as infile:
+            return json.load(infile)
+
+    def _upload_config(self, path: str) -> bool:
+        config = PyTorchModel._load_json(path)
+        response = self._client.ml.put_trained_model(
+            model_id=self.model_id, body=config
+        )
+        return response is not None
+
+    def _upload_vocab(self, path: str) -> bool:
+        vocab = PyTorchModel._load_json(path)
+        return self._client.transport.perform_request(
+            method="PUT",
+            url=f"/_ml/trained_models/{self.model_id}/vocabulary",
+            body=vocab,
+        )
+
+    def _upload_model(
+        self, model_path: str, chunk_size: int = DEFAULT_CHUNK_SIZE
+    ) -> bool:
+        file_stats = os.stat(model_path)
+        total_parts = math.ceil(file_stats.st_size / chunk_size)
+
+        def model_file_chunk_generator():
+            with open(model_path, "rb") as f:
+                while True:
+                    data = f.read(chunk_size)
+                    if not data:
+                        break
+                    yield base64.b64encode(data).decode()
+
+        for i, data in tqdm(
+            enumerate(model_file_chunk_generator(), start=0), total=total_parts
+        ):
+            body = {
+                "total_definition_length": file_stats.st_size,
+                "total_parts": total_parts,
+                "definition": data,
+            }
+            self._client.transport.perform_request(
+                method="PUT",
+                url=f"/_ml/trained_models/{self.model_id}/definition/{i}",
+                body=body,
+            )
+
+        return True
+
+    def upload(
+        self,
+        model_path: str,
```

Comment:
I'm not sure if it makes sense to do this but a lot of times when users are presented with an API that only takes paths to files they immediately as for "can I do it with data in memory"? Does that even make sense here, if so I have an addition review comment to ask if we can support "path or loaded JSON"

Reply:
Yes, definitely. I might need some help making this look right. In the case of model upload, I need a `BinaryIO`. For the configs, it can be easily a path to load or a `Dict` already loaded.